zhaquirks/danfoss/thermostat.py

```diff
@@ -26,113 +55,442 @@
+    """Initiate or Cancel adaptation run."""
+
+    Nothing = 0x00  # not documented in all documentation, but in some places and seems to work
+    Initiate = 0x01
+    Cancel = 0x02
+
+
+class DanfossExerciseDayOfTheWeekEnum(types.enum8):
+    """Day of the week."""
+
+    Sunday = 0
+    Monday = 1
+    Tuesday = 2
+    Wednesday = 3
+    Thursday = 4
+    Friday = 5
+    Saturday = 6
+    Undefined = 7
+
+
+class DanfossOpenWindowDetectionEnum(types.enum8):
+    """Danfoss open window detection judgments."""
+
+    Quarantine = 0x00
+    Closed = 0x01
+    Maybe = 0x02
+    Open = 0x03
+    External = 0x04
+
+
+class DanfossSoftwareErrorCodeBitmap(types.bitmap16):
+    """Danfoss software error code bitmap."""
+
+    Top_pcb_sensor_error = 0x0001
+    Side_pcb_sensor_error = 0x0002
+    Non_volatile_memory_error = 0x0004
+    Unknown_hw_error = 0x0008
+    # 0x0010 = N/A
+    Motor_error = 0x0020
+    # 0x0040 = N/A
+    Invalid_internal_communication = 0x0080
+    # 0x0100 = N/A
+    Invalid_clock_information = 0x0200
+    # 0x0400 = N/A
+    Radio_communication_error = 0x0800
+    Encoder_jammed = 0x1000
+    Low_battery = 0x2000
+    Critical_low_battery = 0x4000
+    # 0x8000 = Reserved
+
+
+class DanfossAdaptationRunStatusBitmap(types.bitmap8):
+    """Danfoss Adaptation run status bitmap."""
+
+    In_progress = 0x0001
+    Valve_characteristic_found = 0x0002
+    Valve_characteristic_lost = 0x0004
+
+
+class DanfossAdaptationRunSettingsBitmap(types.bitmap8):
+    """Danfoss Adaptation run settings bitmap."""
+
+    Disabled = 0x00  # Undocumented, but seems to work
+    Enabled = 0x01
+
+
+class DanfossSetpointCommandEnum(types.enum8):
+    """Set behaviour to change the setpoint."""
+
+    Schedule = 0  # relatively slow
+    User_interaction = 1  # aggressive change
+    Preheat = 2  # invisible to user
+
+
+class DanfossPreheatCommandEnum(types.enum8):
+    """Set behaviour of preheat command.
+    Only one option available, but other values are possible in the future.
+    """
+
+    Force = 0
+
+
+class CustomizedStandardCluster(CustomCluster):
+    """Danfoss customized standard clusters by adding custom attributes.
+
+    Danfoss doesn't allow all standard attributes when manufacturer specific is requested.
+    Therefore, this subclass separates manufacturer specific and standard attributes for Zigbee commands allowing
+    manufacturer specific to be passed for specific attributes, but not for standard attributes.
+    """
+
+    @staticmethod
+    def combine_results(*result_lists):
+        success_global = []
+        failure_global = []
+        for result in result_lists:
+            if len(result) == 1:
+                success_global.extend(result[0])
+            elif len(result) == 2:
+                success_global.extend(result[0])
+                failure_global.extend(result[1])
+
+        if failure_global:
+            return [success_global, failure_global]
+        else:
+            return [success_global]
```

Comment:
Not completely required, but it might be nice if this line could also be covered in the tests.

Reply:
Done.